### PR TITLE
Feature #3036: Window Spooling

### PR DESCRIPTION
**Pull request by hawkfish**
*  Add a Seek method for built on a new PrevScanIndex method to support random access to collections. The scans are linear for now, but that should be good enough for the current task.
* Plumb through everything needed to accumulate WindowDataChunk tuples in a ColumnDataCollection.
* Convert the value storage for the "value" window functions to use collections. Note that the validity masks for IGNORE NULLS and EXCLUDE are still memory resident, but they are a lot smaller.
* Create an explicit class for WindowAggregatorLocalState that can handle building and reading the argument collection in parallel instead of using a DataChunk and locks.
* Convert the naïve aggregator to use collections. 
* Convert the segment tree aggregator to use collections.
* Move all of the WindowDataChunk collection scanning functionality into a single class.
* Convert WindowDistinctAggregator to use paging collections.
* Create a wrapper for multi-threaded appending to WindowDataChunk objects.
* Move RANGE values to a collection.
* Convert the custom window functions to use collections instead of in-memory DataChunks.
* Track the insert data validity and pass it down to avoid checks.
* Pass down context so we can get real memory flushing sizes.

### CODE REVIEW

File: src/common/types/column/column_data_collection.cpp [hunk 1]
```diff
@@ -923,6 +923,28 @@ bool ColumnDataCollection::NextScanIndex(ColumnDataScanState &state, idx_t &chun
...
+		--state.segment_index;
+		state.chunk_index = segments[state.segment_index]->chunk_data.size() + 1;
+		state.current_chunk_state.handles.clear();
+	}
+
+	--state.chunk_index;
+	segment_index = state.segment_index;
+	chunk_index = state.chunk_index - 1;
+	state.next_row_index = state.current_row_index;
+	row_index = state.current_row_index -= segments[state.segment_index]->chunk_data[chunk_index].count;
```
Comment (lnkuiper) [inline src/common/types/column/column_data_collection.cpp:944]: Can we split this assignment up into two lines for readability?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Code changes look good, this is awesome! So `mad`, `mode` and `quantile` can now be computed on partitions if that single partition is larger than memory? Could you add a test for this? What if we have an evil query with multiple like this:
```sql
SELECT mad(c0) OVER (<window with one large partition that doesn't fit in memory>),
       mode(c0) OVER (<same window>),
       quantile(c0, 0.5) OVER (<same window>)
FROM tbl;
```
Could you add a test for that?

**Review by lnkuiper — APPROVED** [review]
Thanks for the changes! The test looks great :)

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**